tests/tape/test_unwrap.py

```diff
@@ -0,0 +1,216 @@
+    ]
+
+    def cost(*p):
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(p[0][0], wires=0)
+            qml.RY(p[1], wires=0)
+            qml.PhaseShift(p[0][1], wires=0)
+            qml.RZ(p[0][2], wires=0)
+
+        with tape.unwrap() as unwrapped_tape:
+            # inside the context manager, all parameters
+            # will be unwrapped to NumPy arrays
+            params = tape.get_parameters(trainable_only=False)
+            assert all(isinstance(i, float) for i in params)
+            assert np.allclose(params, [0.1, 0.2, 0.5, 0.3])
+            assert tape.trainable_params == {0, 2, 3}
+
+        # outside the context, the original parameters have been restored.
+        params = tape.get_parameters(trainable_only=False)
+        assert any(isinstance(i, ArrayBox) for i in params)
+
+        return p[0][0] * p[1] ** 2 * anp.sin(p[0][1]) * anp.exp(-0.5 * p[0][2])
+
+    qml.grad(cost)(*p)
+    qml.jacobian(qml.grad(cost))(*p)
+
+
+def test_unwrap_jax():
+    """Test that unwrapping a tape with JAX parameters
+    works as expected"""
+    jax = pytest.importorskip("jax")
+    from jax import numpy as jnp
+
+    p = [
+        jnp.array(0.1),
+        jnp.array(0.2),
+        0.5,
+        jnp.array(0.3),
+    ]
+
+    with qml.tape.QuantumTape() as tape:
+        qml.RX(p[0], wires=0)
+        qml.RY(p[1], wires=0)
+        qml.PhaseShift(p[2], wires=0)
+        qml.RZ(p[3], wires=0)
+
+    with tape.unwrap() as unwrapped_tape:
+        # inside the context manager, all parameters
+        # will be unwrapped to NumPy arrays
+        params = tape.get_parameters(trainable_only=False)
+        assert all(isinstance(i, float) for i in params)
+        assert np.allclose(params, [0.1, 0.2, 0.5, 0.3])
+
+        # During the forward pass, JAX has no concept of trainable
+        # arrays.
+        assert tape.trainable_params == set()
+
+    # outside the context, the original parameters have been restored.
+    assert tape.get_parameters(trainable_only=False) == p
+
+
+def test_unwrap_jax_backward():
+    """Test that unwrapping a tape with JAX parameters
+    works as expected during a backwards pass"""
+    jax = pytest.importorskip("jax")
+    from jax import numpy as jnp
+    from jaxlib.xla_extension import DeviceArray
+    from jax.interpreters.ad import JVPTracer
+
+    p = [
+        jnp.array([0.1, 0.5, 0.3]),
+        jnp.array(0.2),
+    ]
+
+    def cost(*p):
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(p[0][0], wires=0)
+            qml.RY(p[1], wires=0)
+            qml.PhaseShift(p[0][1], wires=0)
+            qml.RZ(p[0][2], wires=0)
+
+        with tape.unwrap() as unwrapped_tape:
+            # inside the context manager, all parameters
+            # will be unwrapped to NumPy arrays
+            params = tape.get_parameters(trainable_only=False)
+
+            # Currently, JAX does not support converting trainable
+            # parameters to NumPy arrays.

```

Comment:
Interesting! Is that not a problem for us?

Reply:
Your comment above made me think of a solution! So I will modify this test

Reply:
done!